File: VB/CustomDatesPivot/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Windows.Forms

Namespace CustomDatesPivot

    Public Partial Class Form1
        Inherits Form

        Private Table As DataTable

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Shared Function CreatePivotTable(ByVal RowCount As Integer) As DataTable
            Dim rnd As Random = New Random()
            Dim tbl As DataTable = New DataTable()
            tbl.Columns.Add("ID", GetType(Integer))
            tbl.Columns.Add("Name", GetType(String))
            tbl.Columns.Add("Checked", GetType(Boolean))
            tbl.Columns.Add("Number", GetType(Integer))
            tbl.Columns.Add("Value", GetType(Decimal))
            tbl.Columns.Add("AltValue", GetType(Decimal))
            tbl.Columns.Add("Date", GetType(Date))
            For i As Integer = 0 To RowCount - 1
                Dim row As DataRow = tbl.Rows.Add(New Object() {i, String.Format("Name{0}", i Mod 9), i Mod 2 = 0, rnd.Next(10), rnd.Next(100, 1000), rnd.Next(100, 1000), Date.Now.AddDays(rnd.Next(300))})
            Next

            Return tbl
        End Function

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            dateEdit1.DateTime = Date.Now
            dateEdit2.DateTime = Date.Now.AddMonths(12)
            Table = CreatePivotTable(100)
            pivotGridControl1.DataSource = Table
        End Sub

        Private Sub pivotGridControl1_CustomFieldValueCells(ByVal sender As Object, ByVal e As DevExpress.XtraPivotGrid.PivotCustomFieldValueCellsEventArgs)
            Dim index As Integer = pivotGridControl1.GetRowIndex(New Object() {Nothing})
            If index <> -1 Then e.Remove(e.GetCell(False, index))
        End Sub

        Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim DateFields As List(Of String) = New List(Of String)()
            DateFields.Add("Date")
            Dim CustomDates As List(Of Date) = New List(Of Date)()
            Dim [date] As Date = dateEdit1.DateTime
            While [date] < dateEdit2.DateTime
                CustomDates.Add([date])
                Select Case comboBoxEdit1.Text
                    Case "Year"
                        [date] = [date].AddYears(1)
                    Case "Quarter"
                        [date] = [date].AddMonths(3)
                    Case "Month"
                        [date] = [date].AddMonths(1)
                    Case Else
                        [date] = [date].AddDays(1)
                End Select
            End While

            pivotGridControl1.DataSource = New DateDataSourceWrapper(Table.DefaultView, DateFields, CustomDates)
        End Sub
    End Class
End Namespace

File: CS/CustomDatesPivot/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace CustomDatesPivot
{
    public partial class Form1 : Form
    {
        DataTable Table;
        public Form1()
        {
            InitializeComponent();
        }
        private static DataTable CreatePivotTable(int RowCount)
        {
            Random rnd = new Random();

            DataTable tbl = new DataTable();
            tbl.Columns.Add("ID", typeof(int));
            tbl.Columns.Add("Name", typeof(string));
            tbl.Columns.Add("Checked", typeof(bool));
            tbl.Columns.Add("Number", typeof(int));
            tbl.Columns.Add("Value", typeof(decimal));
            tbl.Columns.Add("AltValue", typeof(decimal));
            tbl.Columns.Add("Date", typeof(DateTime));


            for (int i = 0; i < RowCount; i++)
            {
                DataRow row = tbl.Rows.Add(new object[] { i, String.Format("Name{0}", i % 9), i % 2 == 0, rnd.Next(10), rnd.Next(100, 1000), rnd.Next(100, 1000), DateTime.Now.AddDays(rnd.Next(300)) });
            }
            return tbl;
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            dateEdit1.DateTime = DateTime.Now;
            dateEdit2.DateTime = DateTime.Now.AddMonths(12);
            Table = CreatePivotTable(100);
            pivotGridControl1.DataSource = Table;
        }

        private void pivotGridControl1_CustomFieldValueCells(object sender, DevExpress.XtraPivotGrid.PivotCustomFieldValueCellsEventArgs e)
        {
            int index = pivotGridControl1.GetRowIndex(new object[]{null});
            if (index!= -1)
            e.Remove(e.GetCell(false, index));
        }

        private void simpleButton1_Click(object sender, EventArgs e)
        {
            List<string> DateFields = new List<string>();
            DateFields.Add("Date");
            List<DateTime> CustomDates = new  List<DateTime>();
            DateTime date = dateEdit1.DateTime;
            while (date < dateEdit2.DateTime)
            {
                CustomDates.Add(date);
                switch (comboBoxEdit1.Text)
                {
                    case "Year": date = date.AddYears(1);
                        break;
                    case "Quarter": date = date.AddMonths(3);
                        break;
                    case "Month": date = date.AddMonths(1);
                        break;
                    default:
                        date = date.AddDays(1);
                        break;
                }
            }
            pivotGridControl1.DataSource = new DateDataSourceWrapper(Table.DefaultView, DateFields, CustomDates);
        }
    }
}
